### PR TITLE
fixes #10829 - non-admin user cannot update admin password

### PR DESCRIPTION
**Pull request by shlomizadok**
@domcleal -- not sure about the test, the rest seems okay :smile:

### CODE REVIEW

**Comment by domcleal** [comment]
I'd suggest enforcing this at the model level so we don't need to implement this in all of our users controllers.

**Comment by shlomizadok** [comment]
@domcleal - agreed

File: app/models/user.rb [hunk 1]
```diff
@@ -520,6 +521,12 @@ def ensure_default_role
     end
   end
 
+  def ensure_admin_password_changed_by_admin
+    if (self.admin && !User.current.admin) && password_hash_changed?
+      errors.add :password, _("You can't change administrator password")
```
Comment (domcleal) [inline app/models/user.rb]: `cannot be changed by a non-admin user`? (Partial sentences work better with AR field errors I think, and avoids referring in the second person.)

File: test/unit/user_test.rb [hunk 2]
```diff
@@ -312,6 +312,15 @@ def setup_user(operation)
     assert record.save
   end
 
+  test "user cannot set admin password" do
+    setup_user "edit"
+    record = users(:admin)
+    record.password = "123332211"
+    assert_not record.save
```
Comment (domcleal) [inline test/unit/user_test.rb]: superfluous if you're checking validity?

File: app/models/user.rb [hunk 3]
```diff
@@ -308,7 +308,8 @@ def receives?(notification)
   end
 
   def manage_password?
-    auth_source and auth_source.can_set_password?
+    return false if self.admin? && (User.current && !User.current.admin)
```
Comment (domcleal) [inline app/models/user.rb]: `!User.current.try(:admin?)` ?

**Comment by shlomizadok** [comment]
@domcleal - thanks for reviewing and commenting. Fixed

**Comment by dLobatog** [comment]
Merged as e083e965806ebbd121e55d8fa8b210552621a677, thanks @shlomizadok!

**Pull request closed without merging**